eventing/samples/writing-a-source/03-reconcile-sources.md

```diff
@@ -0,0 +1,214 @@
+changes to them.
+
+These edits in the reference project can be viewed at
+https://github.com/grantr/sample-source/pull/5.
+
+## Update Source Status in Reconcile function
+
+The controller's Reconcile function needs to update the reconciled source
+object's Status. Add code to the function to preserve the original object, then
+compare it with the reconciled object and update the status if a change was
+made.
+
+```go
+// Create a copy to determine whether the instance has been modified.
+original := instance.DeepCopy()
+
+// Reconcile the object. If an error occurred, don't return immediately;
+// update the object Status first.
+reconcileErr := r.reconcile(ctx, instance)
+
+// Update object Status if necessary. This happens even if the reconcile
+// returned an error.
+if !equality.Semantic.DeepEqual(original.Status, instance.Status) {
+  log.Info("Updating Status", "request", request.NamespacedName)
+  // An error may occur here if the object was updated since the last Get.
+  // Return the error so the request can be retried later.
+  // This call uses the /status subresource to ensure that the object's spec
+  // is never updated by the controller.
+  if err := r.Status().Update(ctx, instance); err != nil {
+    return reconcile.Result{}, err
+  }
+}
+
+return reconcile.Result{}, reconcileErr
+```
+
+You'll need to define the inner `reconcile` function also. Leave it as a stub
+for now.
+
+```go
+func (r *ReconcileSampleSource) reconcile(ctx context.Context, instance *sourcesv1alpha1.SampleSource) error {
+  return nil
+}
+```
+
+These edits in the reference project can be viewed at
+https://github.com/grantr/sample-source/pull/6.
+
+## Resolve SinkURI from the source's Sink reference
+
+To populate the SinkURI status field, the controller needs to get the object
+named in the Sink reference and look for a URI that can be used as a sink
+address. Many Knative resources have the following fields in their Status:
+
+```yaml
+status:
+  address:
+    hostname: example.default.svc.cluster.local
+```
+
+Resources with these fields are called _Addressable_. The source controller will
+use the value in the `hostname` field to determine a SinkURI.
+
+Add code to the inner `reconcile` function to set the SinkURI field in the
+source Status.
+
+```go
+func (r *ReconcileSampleSource) reconcile(ctx context.Context, instance *sourcesv1alpha1.SampleSource) error {
+
+  // Resolve the Sink URI based on the sink reference.
+  sinkURI, err := r.resolveSinkRef(ctx, instance.Spec.Sink)
+  if err != nil {
+    return fmt.Errorf("Failed to get sink URI: %v", err)
+  }
+
+  // Set the SinkURI field on the SampleSource Status.
+  instance.Status.SinkURI = sinkURI
+
+  //TODO(user): Add additional behavior.
+  return nil
+}
+```
+
+Write the `resolveSinkRef` function. The example below is one example, but there
+are many possibilities. For example, resolving the sink URI of a Kubernetes
+Service would be a useful addition.
```

Comment:
Should this reference a function from eventing-sources as an example?

https://github.com/knative/eventing-sources/blob/master/pkg/controller/sinks/sinks.go#L32 would be the obvious choice.

Reply:
Note that the logging in sinks.go was updated recently to actually print the objects in question, and also uses the `ducks` library from `knative/pkg`.

Reply:
SGTM.